### ISSUE TITLE
Bug in reachability analysis on match with variable-length tuple.

### ISSUE THREAD

**Issue by antcap96** [DetachHead/basedpyright#1704]
### Description

Matching a variable length tuple on an empty list causes all subsequent cases to be marked as unreachable.

Here is an example:

```python
def f()-> tuple[int, ...]:
    return (1,)

x = f()

match x:
    case []:
        print(1)
    case [a]:
        print("unreachable")
    case [a, b]:
        print("unreachable")
    case _:
        print("unreachable")
```
[playground link](https://basedpyright.com/?typeCheckingMode=all&code=CYUwZgBGAUCUC0A%2BCAXArgBwDYgNoEsA7FAGggDpKBdALgCgJGIAnEdZwiaARhNjroAPALxQ4AgLYBDFAGMAFhEH0mEWVIDOICLloNVTDMyIoe-Veq06peg4ePFoAIjSFWUhVIBGOJ%2BaaW2rhSZF62dhBGJs6u7p4%2BIH76jIEQAPoqEVGOLm4gHvLevvxAA)

In the provided pyright playground, it also flags the code as unreachable, but does not raise that as a problem:
[pyright playground link](https://pyright-play.net/?code=CYUwZgBGAUCUC0A%2BCAXArgBwDYgNoEsA7FAGggDpKBdALgCgJGIAnEdZwiaARhNjroAPCAF4ocAQFsAhigDGACwiD6TCHOkBnEBFy0GaphmZEUPfmo3bd0-YaMni0AERpCraYukAjHM4tMVjq40mTedvYQxqYubh5eviD%2BBoxBEAD6qpHRTq7uIJ4KPn78QA)

I could not find an issue about this on pyright.

**Comment by DetachHead** [comment]
very surprising no one seems to have encountered this bug before

**Comment by willwiz** [comment]
I kind of ran into this issue too, when trying out Zed. 

As far as I can find, it should have been fixed in PyRight 1.1.362, see

https://github.com/microsoft/pyright/issues/7826

The weird thing is that this works perfectly fine when using Pylance in VS Code,

```python
from typing import reveal_type


def f(args: tuple[int, ...]) -> int:
    match args:
        case ():
            reveal_type(args)
            # Type of "args" is "tuple[()]"
            return 1
        case (a,):
            reveal_type(args)
            # Type of "args" is "tuple[int]"
            reveal_type(a)
            # Type of "a" is "int"
            return 2
        case (*a,):
            reveal_type(args)
            # Type of "args" is "tuple[int, ...]"
            reveal_type(a)
            # Type of "a" is "list[int]"
            return 3
```

pyright/basedpyright's behavior is also different from mypy and ty, both of which just narrow `args` to `tuple[int, ...]` for all cases. Pyrefly is the only type checker besides Pylance to narrow this correctly, which I suspect is due to its policy for aggressive inference. Maybe part of the fix was bundled into Pylance's typeshed?

Edit: changed thingy to args

**Comment by DetachHead** [comment]
i can't reproduce any different behavior in pylance, but it looks like your example is incomplete because it's missing the definition of `thingy`

**Comment by RBerga06** [comment]
This looks very similar to the upstream issue: https://github.com/microsoft/pyright/issues/11249.

That bug has been introduced in Pyright 1.1.408 (i.e. the latest version as of now); however, latest Pylance release (2025.10.4) is still using Pyright 1.1.407, while Pyright 1.1.408 is included in the latest _prerelease_ (2025.12.101). ([Pylance changelog](https://github.com/microsoft/pylance-release/blob/HEAD/CHANGELOG.md))

@willwiz If this is indeed the same bug, you might try switching to the pre-release version of the extension and see if you can reproduce it.

**Comment by DetachHead** [comment]
thanks, it was probably caused by https://github.com/microsoft/pyright/pull/11085

**Comment by willwiz** [comment]
@RBerga06 Yep, this is it. Switching to pre release on Pylance triggers this as well.

**Comment by willwiz** [comment]
Seems like this was resolved upstream in https://github.com/microsoft/pyright/pull/11370. I'm not observing this in the playgrounds or my code anymore. This issue can probably be closed.

**Comment by DetachHead** [comment]
thanks!
